File: video-generator/src/core/renderer.js
import { exec } from 'child_process';
import path from 'path';
import fs from 'fs';
import { fileURLToPath } from 'url';

// Absolute path to the video-generator project root (two levels up from src/core/)
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const PROJECT_ROOT = path.resolve(__dirname, '..', '..');

/**
 * Stage 6: Manim Renderer
 * Runs the Manim CLI to render a Python script into an MP4 video.
 * Returns the path to the rendered MP4 file.
 */
export async function renderScene(pythonFilePath, sceneClassName, quality = 'l') {
    return new Promise((resolve, reject) => {
        const qualityMap = { l: '480p15', m: '720p30', h: '1080p60', k: '2160p60' };
        const qualityFlag = `-pq${quality}`;
        const outputDir = path.join(PROJECT_ROOT, 'output');
        const scriptName = path.basename(pythonFilePath, '.py');

        console.log(`[Renderer] Rendering ${sceneClassName} at ${qualityMap[quality] || quality}...`);

        const command = `python -m manim ${qualityFlag} "${pythonFilePath}" ${sceneClassName} --media_dir "${outputDir}"`;

        exec(command, (error, stdout, stderr) => {
            if (error) {
                console.error(`[Renderer] Error rendering ${sceneClassName}:`, stderr);
                reject(error);
                return;
            }

            // Resolve the path to the rendered MP4
            const resolution = qualityMap[quality] || '480p15';
            const mp4Path = path.join(outputDir, 'videos', scriptName, resolution, `${sceneClassName}.mp4`);

            console.log(`[Renderer] ✅ Rendered: ${mp4Path}`);
            resolve(mp4Path);
        });
    });
}

File: backend/app/video_generator/renderer.py
import asyncio
import subprocess
import os
import sys
import traceback
from typing import Optional

BACKEND_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', '..'))

async def render_scene(python_file_path: str, scene_class_name: str, quality: str = 'l') -> str:
    quality_map = {'l': '480p15', 'm': '720p30', 'h': '1080p60', 'k': '2160p60'}
    # Use backend/media/videos for outputs
    output_dir = os.path.join(BACKEND_ROOT, 'media', 'manim_output')
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    script_name = os.path.splitext(os.path.basename(python_file_path))[0]
    resolution = quality_map.get(quality, '480p15')
    print(f"[Renderer] Rendering {scene_class_name} at {resolution}...")
    print(f"[Renderer] Python executable: {sys.executable}")
    print(f"[Renderer] File path: {python_file_path}")
    print(f"[Renderer] File exists: {os.path.exists(python_file_path)}")

    # Use sys.executable to ensure we use the same environment
    command = [
        sys.executable, "-m", "manim", 
        "render",
        f"-q{quality}", 
        python_file_path, 
        scene_class_name, 
        "--media_dir", output_dir
    ]
    print(f"[Renderer] Command: {' '.join(command)}")

    try:
        # Use subprocess.run in a thread to avoid Windows NotImplementedError
        # with asyncio.create_subprocess_exec
        result = await asyncio.to_thread(
            subprocess.run,
            command,
            capture_output=True,
            text=True
        )

        stdout_text = result.stdout.strip()
        stderr_text = result.stderr.strip()

        print(f"[Renderer] Return code: {result.returncode}")
        if stdout_text:
            print(f"[Renderer] STDOUT: {stdout_text[:500]}")
        if stderr_text:
            print(f"[Renderer] STDERR: {stderr_text[:500]}")

        if result.returncode != 0:
            combined_error = f"STDOUT: {stdout_text}\nSTDERR: {stderr_text}"
            raise Exception(combined_error)

        mp4_path = os.path.normpath(os.path.join(output_dir, 'videos', script_name, resolution, f"{scene_class_name}.mp4"))

        print(f"[Renderer] Checking for MP4 at: {mp4_path}")
        if os.path.exists(mp4_path):
            print(f"[Renderer] ✅ Found: {mp4_path}")
        else:
            print(f"[Renderer] ❌ File not found at expected path: {mp4_path}")
            parent_dir = os.path.dirname(mp4_path)
            if os.path.exists(parent_dir):
                 try:
                     print(f"[Renderer] Parent dir contains: {os.listdir(parent_dir)}")
                 except Exception:
                     pass
            else:
                 print(f"[Renderer] Parent dir does NOT exist: {parent_dir}")

        return mp4_path
    except Exception as e:
        print(f"[Renderer] 💥 FULL EXCEPTION in render_scene: {type(e).__name__}: {e}")
        print(f"[Renderer] 💥 Traceback:\n{traceback.format_exc()}")
        raise
